### DISCUSSION TITLE
How do I include custom JS properly/elegantly?

### DISCUSSION THREAD

**Discussion by sarkrui** [adityatelange/hugo-PaperMod#491] [Q&A] [answered]
Hello all, I wanted to include [plyr](https://github.com/sampotts/plyr) in my shortcode without referencing from an external site. 

To do so, I placed the necessary `js/css` files in `project/assets/`and referenced `plyr.js` in `layout/shortcodes/plyr.html` by
### Method 1
```html
{{- $plyr := resources.Get "js/plyr.polyfilled.js" | minify | fingerprint}}
<head>
    <script type="application/javascript" src="{{ $plyr.Permalink }}"></script>
</head>
```
### Results
1. It seems there is **no need** to reference the `plyr.css` in the shortcode and hugo will reference and bundle it automatically, whereas as for `plyr.js`, I do need to reference it in the head. **Why is it? was it a configuration on my side?**
2. the minified plyr is exported to the `/public` instead of the one PaperMod is using, `/public/assets/js`, which is ummm not consistent with the project (not elegant). ![image](https://user-images.githubusercontent.com/14228455/128315954-c8c8a3f8-f3bc-4ef4-b74d-aee3f42ba9e7.png) and I wanted the `plyr.min.<fingerprint>` to be placed together with other js. 

To address the above issues, I tried another method, 
### Methods 2
``` html
{{- $blank := resources.Get "js/blank.js" }}
{{- $js := resources.Get "js/plyr.polyfilled.js" }}
{{- $plyr := (slice $blank $js) | resources.Concat "assets/js/plyr.js" | minify | fingerprint}}
<head>
    <script type="application/javascript" src="{{ $plyr.Permalink }}"></script>
</head>
```
I defined a dummy `blank.js`, so I could use `resources.Contact` to export/mount the minified to a designated directory e.g., `public/assets/js`. It worked! 
![image](https://user-images.githubusercontent.com/14228455/128316541-a0d02057-3d54-4a80-8145-0a44333051b0.png)

but I am wondering if there's a bit more elegant way to achieve so. **Thank you**!

**Comment by pixelfabs** [comment, marked as answer]
Answers to your questions and confirmation for your used method:

> It seems there is no need to reference the plyr.css in the shortcode and hugo will reference and bundle it automatically, whereas as for plyr.js, I do need to reference it in the head. Why is it? was it a configuration on my side?

There's no need for referencing CSS files, because Papermod tells Hugo where to find them and what to with it, e.g. find the files in /assets/css/* and bundle it together. This is not something Hugo does, Papermod tells Hugo to do so.

BUT, **PaperMod only bundles CSS files, not JS files**. So, in head.html there's no reference for "bundling" js files. The ones that do get included, but NOT bundled, are defined in head.html (HighLight.js and Fuse.js). 

So, no configuration error on your part. PaperMod doesn't bundle JS files.

And with that, your second method is correct and the only one. Either define it in head.html or if you like JS to be bundled by default, you need to use CSS bundling code and adjust it for JS.

**Reply by sarkrui** [reply]
Hi @pixelfabs  thank you for your answer! Very clear and comprehensive!
